File: lib/geoscript/filter/filter.js
var FILTER_UTIL = require("./util");
var Factory = require("../factory").Factory;
var UTIL = require("../util");
var GeoObject = require("../object").GeoObject;

var _Filter = Packages.org.opengis.filter.Filter;
var CQL = Packages.org.geotools.filter.text.cql2.CQL;
var ECQL = Packages.org.geotools.filter.text.ecql.ECQL;
var Parser = Packages.org.geotools.xml.Parser;
var Encoder = Packages.org.geotools.xml.Encoder;
var _factory = Packages.org.geotools.factory;
var FilterFactory2 = _factory.CommonFactoryFinder.getFilterFactory2(_factory.GeoTools.getDefaultHints());

var OGC = {
    "1.0": {
        OGC: Packages.org.geotools.filter.v1_0.OGC,
        OGCConfiguration: Packages.org.geotools.filter.v1_0.OGCConfiguration
    },
    "1.1": {
        OGC: Packages.org.geotools.filter.v1_1.OGC,
        OGCConfiguration: Packages.org.geotools.filter.v1_1.OGCConfiguration
    }
};


/** api: (define)
 *  module = filter
 *  class = Filter
 */
var Filter = exports.Filter = UTIL.extend(GeoObject, {
    
    /** api: constructor
     *  .. class:: Filter
     *
     *      :arg cql: `String` A CQL string representing filter constraints.
     *
     *      Create a new filter to express constraints.  Filters are typically
     *      used when querying features from a layer.  A feature will be
     *      returned in a query if the filter's :meth:`evaluate` method returns
     *      `true` for the given feature.
     *
     *      Filters are created using Common Query Language (CQL).
     */
    constructor: function Filter(cql) {
        if (cql) {
            if (typeof cql !== "string") {
                cql = cql.cql;
            }
            var _filter;
            try {
                _filter = ECQL.toFilter(cql);
            } catch (err) {
                try {
                    _filter = CQL.toFilter(cql);
                } catch (err2) {
                    throw err;
                }
            }
            this._filter = _filter;
        }
    },
    
    /** api: method[evaluate]
     *  :arg feature: :class:`feature.Feature` A feature.
     *  :returns: ``Boolean``  The feature matches the filter.
     *
     *  Determine whether a feature matches the constraints of the filter.
     */
    evaluate: function(feature) {
        return Boolean(this._filter.evaluate(feature._feature));
    },
    
    /** api: property[not]
     *  :class:`filter.Filter`
     *  A filter that represents the negation of the constraints in this filter.
     */
    get not() {
        return Filter.from_(FilterFactory2.not(this._filter));
    },

    /** api: method[and]
     *  :arg filter: :class:`filter.Filter` Input filter.
     *  :returns: :class:`filter.Filter`
     *  Returns a new filter that is the logical AND of this filter and the 
     *  input filter.  Provide multiple arguments to AND multiple filters.
     */
    and: function(filter) {
        var filters = Array.prototype.slice.call(arguments);
        filters.push(this);
        return Filter.and(filters);
    },
    
    /** api: method[or]
     *  :arg filter: :class:`filter.Filter` Input filter.
     *  :returns: :class:`filter.Filter`
     *  Returns a new filter that is the logical OR of this filter and the 
     *  input filter.  Provide multiple arguments to OR multiple filters.
     */
    or: function(filter) {
        var filters = Array.prototype.slice.call(arguments);
        filters.push(this);
        return Filter.or(filters);
    },
    
    /** api: property[cql]
     *  ``String``
     *  The CQL string that represents constraints in this filter.
     */
    get cql() {
        var string;
        try {
            string = ECQL.toECQL(this._filter);
        } catch (err) {
            string = CQL.toCQL(this._filter);
        }
        return String(string);
    },
    
    get config() {
        return {
            type: "Filter",
            cql: this.cql
        };
    },
    
    /** private: method[toXML]
     *  :arg version: ``String`` Filter Encoding specification version
     *      (default is `"1.0"`).
     *  :arg pretty: ``Boolean`` Use whitespace to indent document elements
     *      (default is `false`).
     *  :returns: ``String``
     *
     *  Generate an XML document string following the Filter Encoding
     *  specification.
     */
    toXML: function(version, pretty) {
        version = version || "1.0";
        var _config = new OGC[version].OGCConfiguration();
        var ogc = OGC[version].OGC.getInstance();
        var encoder = new Encoder(_config);
        encoder.setIndenting(!!pretty);
        encoder.setOmitXMLDeclaration(true);
        var out = new java.io.ByteArrayOutputStream();
        encoder.encode(this._filter, ogc.Filter, out);
        return String(java.lang.String(out.toByteArray()));
    },
    
    toFullString: function() {
        var str;
        try {
            str = this.cql;
        } catch (err) {
            str = String(this._filter);
        }
        return str;
    }
    
});

Filter.from_ = function(_filter) {
    var filter = new Filter();
    filter._filter = _filter;
    return filter;
};

Filter.PASS = Filter.from_(_Filter.INCLUDE);
Filter.FAIL = Filter.from_(_Filter.EXCLUDE);

// logical operators

var getArrayList = function(filters) {
    var len = filters.length;
    var list = new java.util.ArrayList(len);
    var filter;
    for (var i=0; i<len; ++i) {
        filter = filters[i];
        if (!(filter instanceof Filter)) {
            filter = new Filter(filter);
        }
        list.add(filter._filter);
    }
    return list;
};

Filter.and = function(filters) {
    return Filter.from_(FilterFactory2.and(getArrayList(filters)));    
};

Filter.or = function(filters) {
    return Filter.from_(FilterFactory2.or(getArrayList(filters)));
};

Filter.not = function(filter) {
    if (!(filter instanceof Filter)) {
        filter = new Filter(filter);
    }
    return Filter.from_(FilterFactory2.not(filter._filter));
};

Filter.fids = function(fids) {
    var _filter = FilterFactory2.createFidFilter();
    for (var i=0, len=fids.length; i<len; ++i) {
        _filter.addFid(fids[i]);
    }
    return Filter.from_(_filter);
};

/** api: example
 *  Examples of filters that represent various simple constraints:
 * 
 *  .. code-block:: javascript
 * 
 *      js> var namedFoo = new FILTER.Filter("name = 'foo'");
 *      js> var oneThing = new FILTER.Filter("thing = 1");
 *      js> var few = new FILTER.Filter("count < 4");
 *      js> var many = new FILTER.Filter("count > 36");
 *      js> var teens = new FILTER.Filter("age BETWEEN 13 AND 19");
 *
 *  Examples of filters representing spatial constraints:
 *
 *  .. code-block:: javascript
 *  
 *      js> var box = new FILTER.Filter("BBOX(the_geom, -10, -10, 10, 10)");
 *      js> var close = new FILTER.Filter("DWITHIN(the_geom, POINT(1 0), 3, kilometers)");
 *      js> var has = new FILTER.Filter("CONTAINS(the_geom, POINT(1 0))");
 *      js> var hit = new FILTER.Filter("INTERSECTS(the_geom, LINESTRING(0 0, 1 1))");
 */

// register a filter factory for the module
FILTER_UTIL.register(new Factory(Filter, {
    handles: function(config) {
        return true;
    }
}));

File: src/main/java/org/geoscript/js/filter/Filter.java
package org.geoscript.js.filter;

import org.geoscript.js.GeoObject;
import org.geoscript.js.feature.Feature;
import org.geotools.factory.CommonFactoryFinder;
import org.geotools.util.factory.GeoTools;
import org.geotools.filter.text.cql2.CQL;
import org.geotools.filter.text.cql2.CQLException;
import org.geotools.filter.text.ecql.ECQL;
import org.geotools.xsd.Encoder;
import org.mozilla.javascript.*;
import org.mozilla.javascript.annotations.JSConstructor;
import org.mozilla.javascript.annotations.JSFunction;
import org.mozilla.javascript.annotations.JSGetter;
import org.mozilla.javascript.annotations.JSStaticFunction;
import org.opengis.filter.FilterFactory2;
import org.opengis.filter.identity.FeatureId;

import javax.xml.namespace.QName;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class Filter extends GeoObject implements Wrapper {

    /**
     * serialVersionUID
     */
    private static final long serialVersionUID = 4426338466323185386L;

    private org.opengis.filter.Filter filter;

    private static FilterFactory2 factory = CommonFactoryFinder.getFilterFactory2(GeoTools.getDefaultHints());

    /**
     * The Prototype constructor
     */
    public Filter() {
    }

    public Filter(org.opengis.filter.Filter filter) {
        this.filter = filter;
    }

    public Filter(Scriptable scope, org.opengis.filter.Filter filter) {
        this(filter);
        this.setParentScope(scope);
        this.setPrototype(Module.getClassPrototype(Filter.class));
    }

    public Filter(String cql) {
        this(fromCQL(cql));
    }

    public Filter(Scriptable scope, String cql) {
        this(scope, fromCQL(cql));
    }

    /**
     * Create a GeoTools Filter from a CQL String
     */
    private static org.opengis.filter.Filter fromCQL(String cql) {
        org.opengis.filter.Filter filter;
        try {
            filter = ECQL.toFilter(cql);
        } catch (CQLException ex1) {
            try {
                filter = CQL.toFilter(cql);
            } catch (CQLException ex2) {
                throw ScriptRuntime.constructError("Error", "Can't parse Filter CQL Expression: " + ex2.getMessage() +
                        " or ECQL Expression: " + ex1.getMessage());
            }
        }
        return filter;
    }

    @JSConstructor
    public static Object constructor(Context cx, Object[] args, Function ctorObj, boolean isNewExpr) {
        if (args.length != 1) {
            return new Filter();
        }
        Filter filter;
        Object arg = args[0];
        String cql;
        if (arg instanceof String) {
            cql = (String) arg;
        } else if (arg instanceof NativeObject) {
            NativeObject config = (NativeObject) arg;
            cql = config.get("cql", config).toString();
        } else if (arg instanceof Filter) {
            cql = ((Filter) arg).getCql();
        } else {
            throw ScriptRuntime.constructError("Error", "Cannot create filter from provided value: " + Context.toString(ctorObj));
        }
        if (isNewExpr) {
            filter = new Filter(cql);
        } else {
            filter = new Filter(ctorObj.getParentScope(), cql);
        }
        return filter;
    }

    @JSFunction
    public boolean evaluate(Feature feature) {
        return filter.evaluate(feature.unwrap());
    }

    @JSGetter
    public Filter getNot() {
        return new Filter(factory.not(this.filter));
    }

    @JSFunction
    public Filter and(Scriptable filters) {
        if (filters instanceof NativeArray) {
            NativeArray array = (NativeArray) filters;
            array.add(this.filter);
            return Filter.staticAnd(array);
        } else {
            org.opengis.filter.Filter filter = factory.and(this.filter, ((Filter) filters).unwrap());
            return new Filter(getTopLevelScope(filters), filter);
        }
    }

    @JSFunction
    public Filter or(Scriptable filters) {
        if (filters instanceof NativeArray) {
            NativeArray array = (NativeArray) filters;
            array.add(this.filter);
            return Filter.staticOr(array);
        } else {
            org.opengis.filter.Filter filter = factory.or(this.filter, ((Filter) filters).unwrap());
            return new Filter(getTopLevelScope(filters), filter);
        }
    }

    @JSGetter
    public String getCql() {
        return ECQL.toCQL(this.filter);
    }

    @JSGetter("_filter")
    public org.opengis.filter.Filter getFilter() {
        return this.filter;
    }

    @JSGetter
    @Override
    public Scriptable getConfig() {
        Scriptable obj = super.getConfig();
        obj.put("type", obj, "Filter");
        obj.put("text", obj, getCql());
        return obj;
    }

    @JSFunction
    public String toXML(String version, boolean pretty) throws IOException {
        org.geotools.xsd.Encoder encoder;
        QName qname;
        if (version.equalsIgnoreCase("1.1")) {
            qname = org.geotools.filter.v1_1.OGC.getInstance().Filter;
            org.geotools.filter.v1_1.OGCConfiguration config = new org.geotools.filter.v1_1.OGCConfiguration();
            encoder = new Encoder(config);
        } else {
            qname = org.geotools.filter.v1_0.OGC.getInstance().Filter;
            org.geotools.filter.v1_0.OGCConfiguration config = new org.geotools.filter.v1_0.OGCConfiguration();
            encoder = new Encoder(config);
        }
        encoder.setIndenting(pretty);
        encoder.setOmitXMLDeclaration(true);
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        encoder.encode(this.filter, qname, out);
        return new String(out.toByteArray());
    }

    @Override
    public String toFullString() {
        return getCql();
    }

    @Override
    public org.opengis.filter.Filter unwrap() {
        return filter;
    }

    @Override
    public String getClassName() {
        return getClass().getName();
    }

    private static List<org.opengis.filter.Filter> getFilters(NativeArray array) {
        List<org.opengis.filter.Filter> filters = new ArrayList<>();
        for (int i = 0; i < array.size(); i++) {
            Object item = array.get(i);
            Filter filter;
            if (item instanceof Filter) {
                filter = (Filter) item;
            } else if (item instanceof org.opengis.filter.Filter) {
                filter = new Filter(getTopLevelScope(array), (org.opengis.filter.Filter) item);
            } else {
                filter = new Filter(getTopLevelScope(array), array.get(i).toString());
            }
            filters.add(filter.unwrap());
        }
        return filters;
    }

    @JSStaticFunction("from_")
    public static Filter from(Scriptable filterObject) {
        org.opengis.filter.Filter filter = null;
        if (filterObject instanceof Wrapper) {
            Object obj = ((Wrapper) filterObject).unwrap();
            if (obj instanceof org.opengis.filter.Filter) {
                filter = (org.opengis.filter.Filter) obj;
            }
        }
        if (filter == null) {
            throw ScriptRuntime.constructError("Error", "Cannot create filter from " + Context.toString(filterObject));
        }
        return new Filter(getTopLevelScope(filterObject), filter);
    }

    @JSStaticFunction("and")
    public static Filter staticAnd(NativeArray array) {
        List<org.opengis.filter.Filter> filters = getFilters(array);
        return new Filter(getTopLevelScope(array), factory.and(filters));
    }

    @JSStaticFunction("or")
    public static Filter staticOr(NativeArray array) {
        List<org.opengis.filter.Filter> filters = getFilters(array);
        return new Filter(getTopLevelScope(array), factory.or(filters));
    }

    @JSStaticFunction("not")
    public static Filter staticNot(Scriptable obj) {
        Filter filter;
        String cql;
        if (obj instanceof NativeObject) {
            NativeObject config = (NativeObject) obj;
            cql = config.get("cql", config).toString();
        } else if (obj.getDefaultValue(null) instanceof String) {
            cql = obj.getDefaultValue(null).toString();
        } else {
            throw ScriptRuntime.constructError("Error", "Cannot create filter from provided value: " + Context.toString(obj));
        }
        filter = new Filter(cql);
        return new Filter(getTopLevelScope(obj), factory.not(filter.unwrap()));
    }

    @JSStaticFunction
    public static Filter fids(NativeArray array) {
        Set<FeatureId> featureIds = new HashSet<FeatureId>();
        for (int i = 0; i < array.size(); i++) {
            featureIds.add(factory.featureId(array.get(i).toString()));
        }
        return new Filter(getTopLevelScope(array), factory.id(featureIds));
    }

}
